File: neetcode150/trees/leetcode543.py
from typing import Optional

from data_structures.tree import TreeNode


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        max_depth = 0

        def depth(node: Optional[TreeNode]) -> int:
            if not node:
                return 0

            left = depth(node.left)
            right = depth(node.right)

            nonlocal max_depth
            max_depth = max(max_depth, left + right)

            return max(left, right) + 1

        depth(root)

        return max_depth

File: ts_algorithms/neetcode150/trees/leetcode543.ts
import { TreeNode } from "../../data_structures";

function diameterOfBinaryTree(root: TreeNode | null): number {
  let maxLength = 0;

  function recurse(node: TreeNode | null): number {
    if (!node) {
      return 0;
    }

    const left = recurse(node.left);
    const right = recurse(node.right);

    maxLength = Math.max(maxLength, left + right);

    return Math.max(left, right) + 1;
  }

  recurse(root);

  return maxLength;
}

export default diameterOfBinaryTree;
